delocate/fuse.py

```diff
@@ -102,30 +103,41 @@ def fuse_trees(
     from_tree : str or Path-like
         path of tree to fuse from (update from)
     lib_exts : sequence, optional
-        filename extensions for libraries
+        This parameter is deprecated and should be ignored.
+
+    .. versionchanged:: Unreleased
+        Binary files are auto-detected instead of using `lib_exts` to test file
+        suffixes.
     """
+    if lib_exts:
+        warnings.warn(
+            "`lib_exts` parameter will be removed in the future.",
+            FutureWarning,
+            stacklevel=2,
+        )
     for from_dirpath, dirnames, filenames in os.walk(Path(from_tree)):
-        to_dirpath = pjoin(to_tree, relpath(from_dirpath, from_tree))
+        to_dirpath = Path(to_tree, Path(from_dirpath).relative_to(from_tree))
         # Copy any missing directories in to_path
-        for dirname in tuple(dirnames):
-            to_path = pjoin(to_dirpath, dirname)
-            if not exists(to_path):
-                from_path = pjoin(from_dirpath, dirname)
+        for dirname in dirnames.copy():
+            to_path = Path(to_dirpath, dirname)
+            if not to_path.exists():
+                from_path = Path(from_dirpath, dirname)
                 shutil.copytree(from_path, to_path)
                 # If copying, don't further analyze this directory
                 dirnames.remove(dirname)
-        for fname in filenames:
-            root, ext = splitext(fname)
-            from_path = pjoin(from_dirpath, fname)
-            to_path = pjoin(to_dirpath, fname)
-            if not exists(to_path):
+        for filename in filenames:
+            file = Path(filename)
+            from_path = Path(from_dirpath, file)
+            to_path = Path(to_dirpath, file)
+            if not to_path.exists():
                 _copyfile(from_path, to_path)
-            elif cmp_contents(from_path, to_path):
-                pass
-            elif ext in lib_exts:
-                # existing lib that needs fuse
+                continue
+            if cmp_contents(from_path, to_path):
+                continue
+            try:
+                # Try to fuse this file with lipo
```

Comment:
OK - so you're leaving lipo-fuse to detect if this an executable binary on this platform?   Is this safe - I mean - does `lipo -create` guarantee to raise an error if one of the input files is not a mach-o binary?    I did a quick speedcheck - using `file` to check the filetype first for both files nearly doubles the time to do this check, compared to just calling `lipo -create`, but using the `python-magic` library, and parsing the output, adds virtually nothing to the check - and could also give a more informative message.  What do you think?

Reply:
`lipo` returns error codes on non-library files. It will say the file isn't a library on stderr I think. I did some basic probing to make sure that it's at least more reliable than `otool` but I'm not a native MacOS dev. I'm unsure what to do other than to rely on the current tests.

`python-magic` sounds interesting but `fuse_trees` is not a function which does anything with errors other than ignoring them. A list if `lipo` error messages to ignore might be a better idea as I'm more worried about actual library files silently failing to fuse rather than files being falsely detected as libraries. `lipo` doesn't document its error codes very well so you might need to provide me with a list of which stderr strings should be suppressed.

Reply:
Also keep in mind the previous logic of all files with library-like prefixes being unconditionally fused using `lipo`.

Reply:
Happy to work on stderr - but I'm worried that processing stderr might be fragile across versions.   I wonder, should we raise an error when actual libraries fail to fuse, and propagate that error?  I guess we should decide if we want to propagate errors - for example - when the two libraries are the same architecture.

Reply:
I'm unsure what that error would look like. Does `lipo` even fail in these cases?

I personally wouldn't fail here if the architectures are the same, especially if `lipo` allows that case.

Reply:
Lipo does not allow fusing two files with the same architecture, no:

```bash
$ cp ./delocate/tests/data/np-1.24.1_x86_64_random__sfc64.cpython-311-darwin.so foo.so
$ lipo -create delocate/tests/data/np-1.24.1_x86_64_random__sfc64.cpython-311-darwin.so foo.so -output bar.so
fatal error: /Library/Developer/CommandLineTools/usr/bin/lipo: delocate/tests/data/np-1.24.1_x86_64_random__sfc64.cpython-311-darwin.so and foo.so have the same architectures (x86_64) and can't be in the same fat output file
```